File: impls/objpascal/step4_if_fn_do.pas
program Mal;

{$H+} // Use AnsiString

Uses sysutils,
     fgl,
     mal_readline,
     mal_types,
     mal_func,
     reader,
     printer,
     mal_env,
     core;

var
    Repl_Env : TEnv;
    Line     : string;
    I        : longint;
    Key      : string;

// read
function READ(const Str: string) : TMal;
begin
    READ := read_str(Str);
end;

// eval
function EVAL(Ast: TMal; Env: TEnv) : TMal;
var
    Lst    : TMalList;
    Arr    : TMalArray;
    Arr1   : TMalArray;
    A0Sym  : string;
    LetEnv : TEnv;
    FnEnv  : TEnv;
    Cond   : TMal;
    I      : longint;
    Fn     : TMalFunc;
    Args   : TMalArray;
    OldDict, NewDict :  TMalDict;
begin
    Cond := Env.Get('DEBUG-EVAL');
    if (Cond <> nil) and not (Cond is TMalNil) and not (Cond is TMalFalse) then
       WriteLn('EVAL: ' + pr_str(Ast, True));

    if Ast is TMalSymbol then
    begin
        A0Sym := (Ast as TMalSymbol).Val;
        Cond := Env.Get(A0Sym);
        if Cond = nil then
            raise Exception.Create('''' + A0Sym+ ''' not found');
        Exit(Cond);
    end
    else if Ast is TMalVector then
    begin
        Arr := (Ast as TMalVector).Val;
        SetLength(Arr1, Length(Arr));
        for I := 0 to Length(Arr)-1 do
            Arr1[I]:= EVAL(Arr[I], Env);
        Exit(TMalVector.Create(Arr1));
    end
    else if Ast is TMalHashMap then
    begin
        OldDict := (Ast as TMalHashMap).Val;
        NewDict := TMalDict.Create;
        for I := 0 to OldDict.Count-1 do
            NewDict[OldDict.Keys[I]]:= EVAL(OldDict[OldDict.Keys[I]], Env);
        Exit(TMalHashMap.Create(NewDict));
    end
    else if not (Ast is TMalList) then
        Exit(Ast);

    // Apply list
    Lst := (Ast as TMalList);
    Arr := Lst.Val;
    if Length(Arr) = 0 then
        Exit(Ast);
    if Arr[0] is TMalSymbol then
        A0Sym := (Arr[0] as TMalSymbol).Val
    else
        A0Sym := '__<*fn*>__';

    case A0Sym of
    'def!':
        EVAL := Env.Add((Arr[1] as TMalSymbol), EVAL(Arr[2], ENV));
    'let*':
        begin
            LetEnv := TEnv.Create(Env);
            Arr1 := (Arr[1] as TMalList).Val;
            I := 0;
            while I < Length(Arr1) do
            begin
                LetEnv.Add((Arr1[I] as TMalSymbol), EVAL(Arr1[I+1], LetEnv));
                Inc(I,2);
            end;
            EVAL := EVAL(Arr[2], LetEnv);
        end;
    'do':
        begin
            for I := 1 to Length(Arr) - 2 do
                 Cond := EVAL(Arr[I], Env);
           EVAL := EVAL(Arr[Length(Arr)-1], Env);
        end;
    'if':
        begin
            Cond := EVAL(Arr[1], Env);
            if (Cond is TMalNil) or (Cond is TMalFalse) then
                if Length(Arr) > 3 then
                    EVAL := EVAL(Arr[3], Env)
                else
                    EVAL := TMalNil.Create
            else
                EVAL := EVAL(Arr[2], Env);
        end;
    'fn*':
        begin
            EVAL := TMalFunc.Create(Arr[2], Env, (Arr[1] as TMalList))
        end;
    else
        begin
            Cond := EVAL(Arr[0], Env);
            Args := copy(Arr, 1, Length(Arr) - 1);
            if Cond is TMalFunc then
            begin
                Fn := Cond as TMalFunc;
                for I := 0 to Length(Args) - 1 do
                    Args[I]:= EVAL(Args[I], Env);
                if Fn.Ast = nil then
                    EVAL := Fn.Val(Args)
                else
                begin
                    FnEnv := TEnv.Create(Fn.Env, Fn.Params, Args);
                    EVAL := EVAL(Fn.Ast, FnEnv);
                end

            end
            else
                raise Exception.Create('invalid apply');
        end;
    end;
end;

// print
function PRINT(Exp: TMal) : string;
begin
    PRINT := pr_str(Exp, True);
end;

// repl
function REP(Str: string) : string;
begin
    REP := PRINT(EVAL(READ(Str), Repl_Env));
end;

begin
    Repl_Env := TEnv.Create;

    // core.pas: defined using Pascal
    for I := 0 to core.NS.Count-1 do
    begin
        Key := core.NS.Keys[I];
        Repl_Env.Add(TMalSymbol.Create(Key),
                     TMalFunc.Create(core.NS[Key]));
    end;

    // core.mal: defined using language itself
    REP('(def! not (fn* (a) (if a false true)))');

    while True do
    begin
        try
            Line := _readline('user> ');
            if Line = '' then continue;
            WriteLn(REP(Line))
        except
            On E : MalEOF do Halt(0);
            On E : Exception do
            begin
                WriteLn('Error: ' + E.message);
                WriteLn('Backtrace:');
                WriteLn(GetBacktrace(E));
            end;
        end;
    end;
end.

File: impls/es6/step4_if_fn_do.mjs
import rl from './node_readline.mjs'
const readline = rl.readline
import { _list_Q, Vector } from './types.mjs'
import { BlankException, read_str } from './reader.mjs'
import { pr_str } from './printer.mjs'
import { new_env, env_set, env_get } from './env.mjs'
import { core_ns } from './core.mjs'

// read
const READ = str => read_str(str)

// eval
const dbgevalsym = Symbol.for("DEBUG-EVAL")

const EVAL = (ast, env) => {
    if (dbgevalsym in env) {
        const dbgeval = env_get(env, dbgevalsym)
        if (dbgeval !== null && dbgeval !== false) {
            console.log('EVAL:', pr_str(ast, true))
        }
    }

    if (typeof ast === 'symbol') {
        return env_get(env, ast)
    } else if (ast instanceof Vector) {
        return ast.map(x => EVAL(x, env))
    } else if (ast instanceof Map) {
        let new_hm = new Map()
        ast.forEach((v, k) => new_hm.set(k, EVAL(v, env)))
        return new_hm
    } else if (!_list_Q(ast)) {
        return ast
    }

    if (ast.length === 0) { return ast }

    const [a0, a1, a2, a3] = ast
    switch (typeof a0 === 'symbol' ? Symbol.keyFor(a0) : Symbol(':default')) {
        case 'def!':
            return env_set(env, a1, EVAL(a2, env))
        case 'let*':
            let let_env = new_env(env)
            for (let i=0; i < a1.length; i+=2) {
                env_set(let_env, a1[i], EVAL(a1[i+1], let_env))
            }
            return EVAL(a2, let_env)
        case 'do':
            return ast.slice(1).map(x => EVAL(x, env))[ast.length-2]
        case 'if':
            let cond = EVAL(a1, env)
            if (cond === null || cond === false) {
                return typeof a3 !== 'undefined' ? EVAL(a3, env) : null
            } else {
                return EVAL(a2, env)
            }
        case 'fn*':
            return (...args) => EVAL(a2, new_env(env, a1, args))
        default:
            const [f, ...args] = ast.map(x => EVAL(x, env))
            return f(...args)
    }
}

// print
const PRINT = exp => pr_str(exp, true)

// repl
let repl_env = new_env()
const REP = str => PRINT(EVAL(READ(str), repl_env))

// core.EXT: defined using ES6
for (let [k, v] of core_ns) { env_set(repl_env, Symbol.for(k), v) }

// core.mal: defined using language itself
REP('(def! not (fn* (a) (if a false true)))')

while (true) {
    let line = readline('user> ')
    if (line == null) break
    try {
        if (line) { console.log(REP(line)) }
    } catch (exc) {
        if (exc instanceof BlankException) { continue }
        if (exc instanceof Error) { console.warn(exc.stack) }
        else { console.warn(`Error: ${exc}`) }
    }
}
